api/src/main/java/org/apache/iceberg/io/IterableUtils.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.io;
+
+import java.util.Iterator;
+import java.util.function.Predicate;
+import org.apache.iceberg.metrics.MetricsContext.Counter;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Iterables;
+
+public class IterableUtils {
+
+  private IterableUtils() {
+  }
+
+  /**
+   * Counts the number of elements in the given {@link Iterable} by incrementing the {@link Counter}
+   * instance for each {@link Iterator#next()} call.
+   *
+   * @param counter The {@link Counter} instance to increment on each {@link Iterator#next()} call.
+   * @param iterable The underlying {@link Iterator} to count
+   * @param <T> The underlying type to be iterated.
+   * @return An {@link Iterable} that increments the given counter on each {@link Iterator#next()}
+   *     call.
+   */
+  public static <T> Iterable<T> count(Counter<?> counter, Iterable<T> iterable) {
```

Comment:
It would be safer in general if these methods were moved to `CloseableIterable`. Then if the argument is `Iterable` we can wrap it with a noop close and if the argument is `CloseableIterable` we handle it correctly. Adding versions that dont' support close could easily cause trouble with later refactors.

Reply:
Is this somethind we'd want to do as a follow-up? As it would require to change a few places and I didn't want to pollute this PR with semi-related changes

Reply:
I think as long as we're introducing a `count` method, having it work with `CloseableIterable` makes the most sense, rather than doing it in a follow-up.

Reply:
ok I removed `IterableUtils` and moved relevant things to `CloseableIterable`